File: labs/week8/cpp/camShift.cpp
#include <iostream>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/opencv.hpp>
#include <dlib/opencv.h>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing.h>

using namespace std;
using namespace cv;

int main(int argc, char ** argv)
{
  string filename = "../data/videos/face1.mp4";
  if (argc == 2)
  {
    filename = argv[1];
  }
  VideoCapture cap(filename);

  // Check if webcam opens
  if(!cap.isOpened())
  {
    cout << "Error opening video stream or file" << endl;
    return EXIT_FAILURE;
  }

  // Declare the variables for using later
  Mat frame, roiObject, hsvObject, mask, histObject;
  Mat hsv, backProjectImage, frameClone;

  // Read a frame and find the face region using dlib
  for(int i = 0; i< 10; i++)
    cap >> frame;

  dlib::frontal_face_detector faceDetector = dlib::get_frontal_face_detector();
  dlib::cv_image<dlib::bgr_pixel> dlibIm(frame);

  // Detect faces in the image
  std::vector<dlib::rectangle> faceRects = faceDetector(dlibIm);
  Rect currWindow;
  if(faceRects.size() > 0)
  {
    dlib::rectangle bbox = faceRects[0];
    // modify the dlib rect to opencv rect
    currWindow = Rect(
                    (long)bbox.left(),
                    (long)bbox.top() ,
                    (long)bbox.width(),
                    (long)bbox.height()
                    );
  }
  else
  {
    cout << "Face Not Found " << endl;
    cap.release();
    return 0;
  }

  // Create and image with the face region
  frame(currWindow).copyTo(roiObject);
  cvtColor(roiObject, hsvObject, COLOR_BGR2HSV);

  // Get the mask for calculating histogram of the object and also remove noise
  inRange(hsvObject, Scalar(0, 50, 50), Scalar(180, 256, 256), mask);

  // Split the image into channels for finding the histogram
  vector<Mat> channels(3);
  split(hsvObject, channels);
  imshow("Mask", mask);
  imshow("Object", roiObject);

  // Initialize parameters for histogram
  int histSize = 180;
  float range[] = { 0, 179 };
  const float *ranges[] = { range };

  // Find the histogram and normalize it to have values between 0 to 255
  calcHist( &channels[0], 1, 0, mask, histObject, 1, &histSize, ranges, true, false );
  normalize(histObject, histObject, 0, 255, NORM_MINMAX);

  while(1)
  {
    // Read frame
    cap >> frame;
    if( frame.empty() )
      break;

    // Convert to hsv color space
    cvtColor(frame, hsv, COLOR_BGR2HSV);
    split(hsv, channels);

    // find the back projected image with the histogram obtained earlier
    calcBackProject(&channels[0], 1, 0, histObject, backProjectImage, ranges);
    imshow("Back Projected Image", backProjectImage);

    // Compute the new window using CAM shift in the present frame
    RotatedRect rotatedWindow = CamShift(backProjectImage, currWindow, TermCriteria( TermCriteria::EPS | TermCriteria::COUNT, 10, 1 ));

    // Get the rotatedWindow vertices
    Point2f rotatedWindowVertices[4];
    rotatedWindow.points(rotatedWindowVertices);

    frameClone = frame.clone();

    // Display the current window used for mean shift
    rectangle(frameClone, Point(currWindow.x, currWindow.y), Point(currWindow.x + currWindow.width, currWindow.y + currWindow.height), Scalar(255, 0, 0), 2, LINE_AA);

    // Display the rotated rectangle with the orientation information
    for (int i = 0; i < 4; i++)
      line(frameClone, rotatedWindowVertices[i], rotatedWindowVertices[(i+1)%4], Scalar(0,255,0), 2, LINE_AA);
    imshow("CAMShift Object Tracking Demo", frameClone);

    int k = cv::waitKey(10);
    if (k == 27)
      break;
  }

  cap.release();
  destroyAllWindows();
  return 0;
}

File: labs/week8/python/camShift.py
#!/usr/bin/python
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
# 
# This code is made available to the students of 
# the online course titled "Computer Vision for Faces" 
# by Satya Mallick for personal non-commercial use. 
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC. 
#
# For licensing and other inquiries, please email 
# spmallick@bigvisionllc.com 
# 
import numpy as np
import cv2,sys,dlib,time

if __name__ == '__main__':

  filename = "../data/videos/face1.mp4"
  if len(sys.argv) == 2:
    filename = sys.argv[1]
  cap = cv2.VideoCapture(filename)

  # Read a frame and find the face region using dlib
  for i in range(5):
    ret,frame = cap.read()

  # Detect faces in the image
  faceDetector = dlib.get_frontal_face_detector()
  faceRects = faceDetector(frame, 0)

  if not len(faceRects):
    print("face not found")
    sys.exit()

  bbox = faceRects[0]

  # modify the dlib rect to opencv rect 
  currWindow = (int(bbox.left()), int(bbox.top()), int(bbox.right() - bbox.left()), int(bbox.bottom() - bbox.top()))

  # get the face region from the frame
  roiObject = frame[currWindow[1]:currWindow[1] + currWindow[3], currWindow[0]:currWindow[0] + currWindow[2]]
  hsvObject =  cv2.cvtColor(roiObject, cv2.COLOR_BGR2HSV)

  # GGGet the mask for calculating histogram of the object and also remove noise
  mask = cv2.inRange(hsvObject, np.array((0., 50., 50.)), np.array((180.,255.,255.)))
  cv2.imshow("mask",mask)
  cv2.imshow("Object",roiObject)

  # Find the histogram and normalize it to have values between 0 to 255
  histObject = cv2.calcHist([hsvObject], [0], mask, [180], [0,180])           
  cv2.normalize(histObject, histObject, 0, 255, cv2.NORM_MINMAX)

  # Setup the termination criteria, either 10 iterations or move by atleast 1 pt
  term_crit = ( cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 10, 1 )
  i=0
  while(1):
    ret, frame = cap.read()

    if ret == True:
      # Convert to hsv color space
      hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

      # find the back projected image with the histogram obtained earlier
      backProjectImage = cv2.calcBackProject([hsv], [0], histObject, [0,180], 1)
      cv2.imshow("Back Projected Image", backProjectImage)

      # Compute the new window using CAM shift in the present frame
      rotatedWindow, currWindow = cv2.CamShift(backProjectImage, currWindow, term_crit)

      # Get the window used by mean shift
      x,y,w,h = currWindow
      
      # Get the rotatedWindow vertices
      rotatedWindow = cv2.boxPoints(rotatedWindow)
      rotatedWindow = np.int0(rotatedWindow)
      frameClone = frame.copy()

      # Display the current window used for mean shift
      cv2.rectangle(frameClone, (x,y), (x+w,y+h), (255, 0, 0), 2, cv2.LINE_AA)

      # Display the rotated rectangle with the orientation information
      frameClone = cv2.polylines(frameClone, [rotatedWindow], True, (0, 255, 0), 2, cv2.LINE_AA)
      cv2.putText(frameClone, "{},{},{},{}".format(x,y,w,h), (10, 30), cv2.FONT_HERSHEY_SIMPLEX, .8, (0, 0, 255), 2, cv2.LINE_AA)
      cv2.putText(frameClone, "{}".format(rotatedWindow), (10, 80), cv2.FONT_HERSHEY_SIMPLEX, .8, (0, 0, 255), 2, cv2.LINE_AA)
      cv2.imshow('CAM Shift Object Tracking Demo',frameClone)
      cv2.imwrite('{:03d}.jpg'.format(i),frameClone)
      i+=1

      k = cv2.waitKey(10) & 0xff
      if k == 27:
        break
    else:
      break

  cv2.destroyAllWindows()
  cap.release()
